lib/libutils/ext/mempool.c

```diff
@@ -74,10 +74,15 @@ struct mempool {
 static void get_pool(struct mempool *pool __maybe_unused)
 {
 #if defined(__KERNEL__)
-	if (refcount_inc(&pool->refc)) {
-		if (pool->owner == thread_get_id())
-			return;
-		refcount_dec(&pool->refc);
+	/*
+	 * Owner matches our thread it cannot be changed. If it doesn't
+	 * match it can change any at time we're not holding the mutex to
+	 * any value but our thread id.
+	 */
+	if (atomic_load_int(&pool->owner) == thread_get_id()) {
+		if (!refcount_inc(&pool->refc))
```

Comment:
I think, that this code is still prone to race. You have two atomic operations there, but you need only one. 
Suppose that Thread 0 is at line 107 and Tread 1 is at L82. Thread 1 passes check and thinks that it owns the pool. But Thread 0 resets pool owner.

I can't see how this can be fixed with atomic variables. Basically, you need to execute operation "check owner and  increase refcount" atomically. But you are are doing two atomic operations instead - "check owner" and "increase refcount".

I propose to drop atomics at all and introduce spinlock. With spinlock you really can operate atomically on owner and refcount.

Reply:
Only the thread owning the pool may set it to free. The case you're describing is not supposed to happen, it's like unlocking a mutex with wrong thread. The code doesn't depend on `refcount_inc()` and `refcount_dec()` being atomic, it just happens to be a convenient interface.

Reply:
Ah yes, you are right.